static/src/javascripts/projects/common/modules/crosswords/anagram-helper/clue-preview.js

```diff
@@ -0,0 +1,98 @@
+// @flow
+import React from 'react/addons';
+
+const CluePreview = React.createClass({
+    /**
+     * Get the entries for the preview cells: first filter the user's input to
+     * remove anything anything that's already been entered into the grid.
+     *
+     * With that, we map over the entries, and wherever there's an empty space
+     * we insert one of the shuffled input characters.
+     *
+     * If the user hasn't yet clicked 'shuffle' (this.props.hasShuffled) just
+     * display the entries as they are, preserving any blank spaces.
+     */
+    getEntries(): Array<Object> {
+        const unsolved = this.props.letters.filter(l => !l.entered);
+
+        return this.props.entries.map(entry => {
+            entry.solved = !!entry.value;
+
+            return this.props.hasShuffled
+                ? (entry.value && entry) || unsolved.shift()
+                : entry;
+        });
+    },
+
+    // Checks a object in the form{",":[4,7]}
+    checkIfLetterHasSeparator(locations: Object, letterIndex: number): string {
```

Comment:
Could probably type up the `locations` object

```
type locations {
  ,?: Array<number>
  -?: Array<number>
}
```

Syntax might need checking!